src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -383,249 +507,339 @@ private void TrainModels(Float[][] columnData, int[] rowCounts, IChannel ch)
+                Host.Assert(0 <= iinfo && iinfo < _parent.ColumnPairs.Length);
+                int src = _cols[iinfo];
+                Host.Assert(input.IsColumnActive(src));
+                return input.GetGetter<T>(src);
+            }
 
-                int offs = 0;
-                for (int i = 0; i < cdst; i++)
+            private static void FillValues(float[] model, ref VBuffer<float> src, ref VBuffer<float> dst, int cdst)
+            {
+                int count = src.Count;
+                int length = src.Length;
+                var values = src.Values;
+                var indices = src.Indices;
+                Contracts.Assert(Utils.Size(values) >= count);
+
+                // Since the whitening process produces dense vector, always use dense representation of dst.
+                var a = Utils.Size(dst.Values) >= cdst ? dst.Values : new float[cdst];
+                if (src.IsDense)
                 {
-                    a[i] = DotProduct(model, offs, values, indices, count);
-                    offs += length;
+                    Mkl.Gemv(Mkl.Layout.RowMajor, Mkl.Transpose.NoTrans, cdst, length,
+                        1, model, length, values, 1, 0, a, 1);
                 }
+                else
+                {
+                    Contracts.Assert(Utils.Size(indices) >= count);
+
+                    int offs = 0;
+                    for (int i = 0; i < cdst; i++)
+                    {
+                        a[i] = DotProduct(model, offs, values, indices, count);
+                        offs += length;
+                    }
+                }
+                dst = new VBuffer<float>(cdst, a, dst.Indices);
+            }
+
+            /// <summary>
+            /// Returns a dot product of dense vector 'a' starting from offset 'aOffset' and sparse vector 'b'
+            /// with first 'count' valid elements and their corresponding 'indices'.
+            /// </summary>
+            private static float DotProduct(float[] a, int aOffset, float[] b, int[] indices, int count)
+            {
+                Contracts.Assert(count <= indices.Length);
+                return CpuMathUtils.DotProductSparse(a.AsSpan(aOffset), b, indices, count);
+
             }
-            dst = new VBuffer<Float>(cdst, a, dst.Indices);
         }
+    }
 
-        /// <summary>
-        /// Returns a dot product of dense vector 'a' starting from offset 'aOffset' and sparse vector 'b'
-        /// with first 'count' valid elements and their corresponding 'indices'.
-        /// </summary>
-        private static Float DotProduct(Float[] a, int aOffset, Float[] b, int[] indices, int count)
-        {
-            Contracts.Assert(count <= indices.Length);
-            return CpuMathUtils.DotProductSparse(a.AsSpan(aOffset), b, indices, count);
+    /// <include file='doc.xml' path='doc/members/member[@name="Whitening"]/*'/>
+    public sealed class WhiteningEstimator : IEstimator<WhiteningTransform>
```

Comment:
>IEstimator [](start = 45, length = 10)

Can this derive from TrivialEstimator?

Reply:
My understanding is that it can't. A trivial estimator needs the transform as a parameter to the constructor. However, in case of trainable transforms, you can't provide that until you have access to the data. Therefore, the solution from my understanding is to initialize the estimator by keeping the arguments, and initialize the trained transform only when .fit(data) is called.  This is what I wrote in the checklist that I sent out for transform conversion which Pete proofread.

Reply:
No it can't as it is a trainable transform. 

---
In reply to: [229555723](https://github.com/dotnet/machinelearning/pull/1452#discussion_r229555723) [](ancestors = 229555723)